alpenhorn/auto_import.py

```diff
@@ -1,323 +1,375 @@
-        the node root.
+    node : UpdateableNode
+        The node we're importing onto
+    queue : FairMultiFIFOQueue
+        The tasks queue
+    path : pathlib.PurePath
+        The path we're trying to import
     """
 
-    log.debug('Considering "%s" for import.', file_path)
-
     # Occasionally the watchdog sends events on the node root directory itself. Skip these.
-    if file_path == node.root:
-        log.debug('Skipping import request on the node root itself "%s"', node.root)
+    if path == pathlib.PurePath(node.db.root):
+        log.debug("Skipping import request of node root")
         return
 
-    # Ensure the path is an absolute path within the node
-    if os.path.isabs(file_path):
-        if not in_directory(file_path, node.root):
-            log.error(
-                'File "%s" was not an absolute path within the node "%s"',
-                file_path,
-                node.root,
+    # Strip node root, if path is absolute
+    if path.is_absolute():
+        try:
+            path = path.relative_to(node.db.root)
+        except ValueError:
+            # Not rooted under node.root
+            log.warning(
+                f"skipping import of {path}: not rooted under node {node.db.root}"
             )
             return
-    else:
-        file_path = os.path.join(node.root, file_path)
-    abspath = os.path.normpath(file_path)
 
-    # Skip requests to import a directory. Again these are occasionally sent by the watchdog
-    if os.path.isdir(file_path):
-        log.debug('Path to import "%s" is a directory. Skipping...', file_path)
+    # Ignore the node info file.  NB: this happens even on nodes which
+    # don't use an ALPENHORN_NODE file, meaning even on those nodes,
+    # this path is disallowed as a data file.
+    if str(path) == "ALPENHORN_NODE":
+        log.debug("ignoring ALPENHORN_NODE file during import")
         return
 
-    relpath = os.path.relpath(abspath, node.root)
+    # New import task.  It's better to do this in a worker
+    # because the worker pool will take care of DB connection loss.
+    Task(
+        func=_import_file,
+        queue=queue,
+        key=node.name,
+        args=(node, path),
+        name=f"Import {path} on {node.name}",
+        # If the job fails due to DB connection loss, re-start the
+        # task because unlike tasks made in the main loop, we're
+        # never going to revisit this.
+        requeue=True,
+    )
+
+
+def _import_file(task: Task, node: StorageNode, path: pathlib.PurePath) -> None:
+    """Import `path` on `node` into the DB.  This is run by a worker.
+
+    Parameters
+    ----------
+    task : task.Task
+        The task running this function
+    node : storage.StorageNode
+        The node we are processing.
+    path : pathlib.PurePath
+        The path should be relative to `node.root`
+    """
+
+    log.debug(f'Considering "{path}" for import to node {node.name}.')
```

Comment:
This is not needed.  The worker has just reported that it started task "Import {path} on {node.name}".